### ISSUE TITLE
Apply black to Tests/

### ISSUE THREAD

**Issue by peterjc** [biopython/biopython#2552]
This is a spin out from #2008, where we have agreed to adopt the black code formatting tool for Biopython:

https://github.com/psf/black

That issue focused on applying black to the main code base (``Bio/``, ``BioSQL/``, and also ``Scripts/`` and ``Doc/examples/``).

This issue is for applying ``black`` to the Biopython test suite.

In principle this is a good issue for new contributors. However, many of the tests include expected data values, so compared to the Biopython main code there are more likely to be cases where black does a bad job - which will need some discussion (e.g. can it be improved with extra brackets, or do we turn off black for a portion of a test file?).

You will need to install black, flake8, and associated plugins as described in https://github.com/biopython/biopython/blob/master/CONTRIBUTING.rst - and we recommend turning on the git pre-commit hook).

Example workflow:

```
git checkout master # and make sure it is up to date
git checkout -b black_test_XXX # start a new branch
black Tests/test_XXX.py  # do the conversion
flake8 Tests/test_XXX.py # check for string concatenation etc
emacs Tests/test_XXX.py  # review and make any fixes
git commit Tests/test_XXX.py -m "Apply black to Tests/test_XXX.py"
git push my_fork black_text_XXX  # push the branch to your personal GitHub
```

To avoid duplication of effort, please comment here if you are going to work on a particular file, or group of files.

**Comment by svalqui** [comment]
I'll work on test_PDB*

**Comment by svalqui** [comment]
I'll work on test_SeqIO*

**Comment by svalqui** [comment]
I'll work on test_SearchIO* files.

**Comment by svalqui** [comment]
I'll work on test_pro* files.

**Comment by svalqui** [comment]
I'll work on test_Phylo* files.

**Comment by svalqui** [comment]
I'll work on test_PAML* files.

**Comment by svalqui** [comment]
I'll work on test_T* files.

**Comment by svalqui** [comment]
I'll skip test_Tutorial.py as is raising errors with some doctest.
FAIL: efetch (Bio.Entrez)
FAIL: pairwise2 (Bio)

**Comment by svalqui** [comment]
I'll work on SCOP files

**Comment by peterjc** [comment]
You can try ``python run_tests.py test_Tutoturial.py --offline`` but it sounds like four of the online Entrez tests in the Tutorial were failing. Please confirm if that was only after running black or not. If not, please open a new issue. Thanks.

**Comment by svalqui** [comment]
It was before running black, I have logged an issue for that.

**Comment by svalqui** [comment]
I'll work on test_PopGen* files

**Comment by svalqui** [comment]
I'll work on test_phenotype* files.

**Comment by svalqui** [comment]
I'll work on test_NCBI* files.

**Comment by svalqui** [comment]
test_NCBI_qblast.py is failing, before and after black, so not black related, I'll apply black style and then log an issue.

**Comment by svalqui** [comment]
I'll work on test_m* files.

**Comment by svalqui** [comment]
I'll work on test_l* files.

**Comment by svalqui** [comment]
I'll work on test_k* files.

**Comment by svalqui** [comment]
I'll work on test_HMM* files.

**Comment by svalqui** [comment]
I'll work on test_g* files.

**Comment by svalqui** [comment]
I'll work on test_Entrez files.

**Comment by svalqui** [comment]
I'll work on test_Emboss files.

**Comment by svalqui** [comment]
I'll work on test_AlignIO files.

**Comment by svalqui** [comment]
All that could had been grouped had black style applied, now looking at individual files.

**Comment by gandalfthegrape** [comment]
Hello, I'm still new to contributing. Unless I'm mistaken, it looks like test_Nexus.py has not been updated yet. I'm a little confused as to what to do though. When I run black and then "flake8 test_Nexus.py" it only shows 4 lines with trailing whitespace, but when I fix that and go to commit, the pre-commit hook complains about all the missing docstrings for public classes and functions. Glancing at some of the other users work, it looks like some of them also have classes with missing docstrings, eg. class TestSwissProt in test_SwissProt.py.

This file also has the issue of black wanting to expand all the number lists to many lines, but I'll worry about that later.

Any guidance is appreciated!

**Comment by MarkusPiotrowski** [comment]
Are you using the `.flake8` config file of the master branch? E.g. the errors you mentioned are, afaik, turned off for the `test` folder.

**Comment by gandalfthegrape** [comment]
I'm not quite sure. When I run `flake8 Tests/test_Nexus.py` it does not show errors, but when I go to commit it gives me several errors that the `.flake8` file should be ignoring.

**Comment by svalqui** [comment]
The file .flake8 has a section where list some errors that can be ignored:
https://github.com/biopython/biopython/blob/6c75d2da18afdc89b6b3ef994d87c601f856ae27/.flake8#L45-48
for example if you run flake and get this:
`test_Nexus.py:1168:1: D102 Missing docstring in public method`
you can ignore it, that relates to missing documentation, a missing docstring in a method not this issue.

if you get this:
`test_Nexus.py:727:46: W291 trailing whitespace`
then you should fix it, this is not been ignored, most likely a space in the wrong place that makes the code not PEP compliant or looks odd.

This issue relates to the styling of the code only, black makes the code look even, however in this project the results of black styling is not necessary better, for example some list are presented as one item per line making long lists hard to read, that might need adjustment, core members or contributors will advice accordingly.

I hope this helps.

**Comment by gandalfthegrape** [comment]
Thanks you for the advice!  That makes sense, I couldn't figure out how to force a commit when it gives those errors I mentioned, but I realized I can use the `--no-verify` argument when I'm sure those are the only errors. I'll get working on this then.

**Comment by MarkusPiotrowski** [comment]
Probably, your pre-commit hook is set wrongly. Have you tried to remove and re-set it? Are you in the `\biopython` folder when doing the commit? I mean, the place from where you type `git commit ...`

**Comment by peterjc** [comment]
One simple possibility: It could be you had a flake8 error, did the git add, git commit - and got the warning. If you then fixed the flake8 error, and repeated git commit without doing git add first, it would still be trying to commit the bad version, not the newer fixed version. If so, do another git add, then retry the git commit.

**Comment by peterjc** [comment]
Many thanks to @svalqui - just merged a bunch of their pull requests for this issue which had been left pending.

**Comment by martonlanga** [comment]
Went through the files, to see which ones need black. 120/210 already done!
Feel free to tag me if you make a PR, and I'll edit this to make it up to date.

-   [ ] `common_BioSQL.py`
-   [ ] `common_BioSQL_online.py`
-   [ ] `gen_results.py`
-   [x] `pairwise2_testCases.py`
-   [ ] `requires_internet.py`
-   [ ] `requires_wise.py`
-   [x] `run_tests.py`
-   [ ] `search_tests_common.py`
-   [ ] `seq_tests_common.py`
-   [ ] `test_Ace.py`
-   [ ] `test_Affy.py`
-   [x] `test_AlignIO.py`
-   [x] `test_AlignIO_ClustalIO.py`
-   [x] `test_AlignIO_EmbossIO.py`
-   [x] `test_AlignIO_FastaIO.py`
-   [x] `test_AlignIO_MauveIO.py`
-   [x] `test_AlignIO_PhylipIO.py`
-   [x] `test_AlignIO_convert.py`
-   [ ] `test_AlignInfo.py`
-   [ ] `test_Application.py`
-   [ ] `test_BWA_tool.py`
-   [ ] `test_BioSQL_MySQLdb.py`
-   [x] `test_BioSQL_MySQLdb_online.py`
-   [ ] `test_BioSQL_mysql_connector.py`
-   [x] `test_BioSQL_mysql_connector_online.py`
-   [ ] `test_BioSQL_psycopg2.py`
-   [x] `test_BioSQL_psycopg2_online.py`
-   [x] `test_BioSQL_sqlite3.py`
-   [x] `test_BioSQL_sqlite3_online.py`
-   [ ] `test_Blast_Record.py`
-   [ ] `test_CAPS.py`
-   [ ] `test_Chi2.py`
-   [ ] `test_ClustalOmega_tool.py`
-   [ ] `test_Clustalw_tool.py`
-   [ ] `test_Cluster.py`
-   [ ] `test_CodonTable.py`
-   [ ] `test_ColorSpiral.py`
-   [ ] `test_Compass.py`
-   [ ] `test_Consensus.py`
-   [ ] `test_Crystal.py`
-   [ ] `test_DSSP_tool.py`
-   [ ] `test_Dialign_tool.py`
-   [ ] `test_EMBL_unittest.py`
-   [x] `test_Emboss.py`
-   [x] `test_EmbossPhylipNew.py`
-   [x] `test_EmbossPrimer.py`
-   [x] `test_Entrez.py`
-   [x] `test_Entrez_online.py`
-   [x] `test_Entrez_parser.py`
-   [ ] `test_Enzyme.py`
-   [ ] `test_ExPASy.py`
-   [ ] `test_FSSP.py`
-   [ ] `test_Fasttree_tool.py`
-   [ ] `test_File.py`
-   [x] `test_GenBank.py`
-   [x] `test_GenomeDiagram.py`
-   [x] `test_GraphicsBitmaps.py`
-   [x] `test_GraphicsChromosome.py`
-   [x] `test_GraphicsDistribution.py`
-   [x] `test_GraphicsGeneral.py`
-   [x] `test_HMMCasino.py`
-   [x] `test_HMMGeneral.py`
-   [ ] `test_KEGG.py`
-   [ ] `test_KEGG_online.py`
-   [ ] `test_KGML_graphics.py`
-   [ ] `test_KGML_graphics_online.py`
-   [ ] `test_KGML_nographics.py`
-   [ ] `test_KeyWList.py`
-   [x] `test_LogisticRegression.py`
-   [ ] `test_MSAProbs_tool.py`
-   [ ] `test_MafIO_index.py`
-   [ ] `test_Mafft_tool.py`
-   [ ] `test_MarkovModel.py`
-   [ ] `test_Medline.py`
-   [ ] `test_Muscle_tool.py`
-   [ ] `test_NACCESS_tool.py`
-   [ ] `test_NCBITextParser.py`
-   [ ] `test_NCBIXML.py`
-   [x] `test_NCBI_BLAST_tools.py`
-   [x] `test_NCBI_qblast.py`
-   [ ] `test_NMR.py`
-   [ ] `test_NaiveBayes.py`
-   [ ] `test_Nexus.py`
-   [x] `test_PAML_baseml.py`
-   [x] `test_PAML_codeml.py`
-   [x] `test_PAML_tools.py`
-   [x] `test_PAML_yn00.py`
-   [ ] `test_PDB.py`
-   [x] `test_PDBList.py`
-   [ ] `test_PDB_Dice.py`
-   [x] `test_PDB_FragmentMapper.py`
-   [x] `test_PDB_KDTree.py`
-   [x] `test_PDB_MMCIF2Dict.py`
-   [ ] `test_PDB_MMCIFParser.py`
-   [x] `test_PDB_Polypetide.py`
-   [x] `test_PDB_ResidueDepth.py`
-   [ ] `test_PDB_StructureAlignment.py`
-   [x] `test_PDB_Superimposer.py`
-   [ ] `test_PDB_parse_pdb_header.py`
-   [x] `test_PDB_vectors.py`
-   [ ] `test_PQR.py`
-   [ ] `test_Pathway.py`
-   [ ] `test_Phd.py`
-   [x] `test_Phylo.py`
-   [x] `test_PhyloXML.py`
-   [x] `test_Phylo_CDAO.py`
-   [x] `test_Phylo_NeXML.py`
-   [x] `test_Phylo_matplotlib.py`
-   [x] `test_Phylo_networkx.py`
-   [x] `test_PopGen_GenePop.py`
-   [x] `test_PopGen_GenePop_EasyController.py`
-   [x] `test_PopGen_GenePop_nodepend.py`
-   [ ] `test_Prank_tool.py`
-   [x] `test_Probcons_tool.py`
-   [x] `test_ProtParam.py`
-   [ ] `test_QCPSuperimposer.py`
-   [x] `test_RCSBFormats.py`
-   [x] `test_Restriction.py`
-   [x] `test_SCOP_Astral.py`
-   [x] `test_SCOP_Cla.py`
-   [x] `test_SCOP_Des.py`
-   [ ] `test_SCOP_Dom.py`
-   [x] `test_SCOP_Hie.py`
-   [x] `test_SCOP_Raf.py`
-   [x] `test_SCOP_Residues.py`
-   [x] `test_SCOP_Scop.py`
-   [x] `test_SCOP_online.py`
-   [x] `test_SVDSuperimposer.py`
-   [x] `test_SearchIO_blast_tab.py`
-   [ ] `test_SearchIO_blast_tab_index.py`
-   [x] `test_SearchIO_blast_text.py`
-   [x] `test_SearchIO_blast_xml.py`
-   [ ] `test_SearchIO_blast_xml_index.py`
-   [x] `test_SearchIO_blat_psl.py`
-   [ ] `test_SearchIO_blat_psl_index.py`
-   [x] `test_SearchIO_exonerate.py`
-   [ ] `test_SearchIO_exonerate_text_index.py`
-   [ ] `test_SearchIO_exonerate_vulgar_index.py`
-   [x] `test_SearchIO_fasta_m10.py`
-   [ ] `test_SearchIO_fasta_m10_index.py`
-   [x] `test_SearchIO_hhsuite2_text.py`
-   [x] `test_SearchIO_hmmer2_text.py`
-   [ ] `test_SearchIO_hmmer2_text_index.py`
-   [x] `test_SearchIO_hmmer3_domtab.py`
-   [x] `test_SearchIO_hmmer3_domtab_index.py`
-   [x] `test_SearchIO_hmmer3_tab.py`
-   [ ] `test_SearchIO_hmmer3_tab_index.py`
-   [x] `test_SearchIO_hmmer3_text.py`
-   [x] `test_SearchIO_hmmer3_text_index.py`
-   [x] `test_SearchIO_interproscan_xml.py`
-   [x] `test_SearchIO_legacy.py`
-   [x] `test_SearchIO_model.py`
-   [x] `test_SearchIO_write.py`
-   [x] `test_SeqFeature.py`
-   [x] `test_SeqIO.py`
-   [x] `test_SeqIO_AbiIO.py`
-   [x] `test_SeqIO_FastaIO.py`
-   [x] `test_SeqIO_Gck.py`
-   [x] `test_SeqIO_Insdc.py`
-   [x] `test_SeqIO_NibIO.py`
-   [ ] `test_SeqIO_PdbIO.py`
-   [x] `test_SeqIO_QualityIO.py`
-   [ ] `test_SeqIO_SeqXML.py`
-   [x] `test_SeqIO_SnapGene.py`
-   [ ] `test_SeqIO_Xdna.py`
-   [x] `test_SeqIO_features.py`
-   [x] `test_SeqIO_index.py`
-   [x] `test_SeqIO_online.py`
-   [ ] `test_SeqIO_write.py`
-   [x] `test_SeqRecord.py`
-   [x] `test_SeqUtils.py`
-   [x] `test_Seq_objs.py`
-   [x] `test_SffIO.py`
-   [x] `test_SubsMat.py`
-   [x] `test_SwissProt.py`
-   [x] `test_TCoffee_tool.py`
-   [x] `test_TogoWS.py`
-   [x] `test_TreeConstruction.py`
-   [x] `test_Tutorial.py`
-   [x] `test_UniGene.py`
-   [x] `test_UniProt_GOA.py`
-   [ ] `test_Uniprot.py`
-   [x] `test_Wise.py`
-   [x] `test_XXmotif_tool.py`
-   [ ] `test_align.py`
-   [ ] `test_align_substitution_matrices.py`
-   [ ] `test_bgzf.py`
-   [ ] `test_cellosaurus.py`
-   [ ] `test_codonalign.py`
-   [ ] `test_geo.py`
-   [ ] `test_kNN.py`
-   [x] `test_lowess.py`
-   [ ] `test_mmtf.py`
-   [ ] `test_mmtf_online.py`
-   [ ] `test_motifs.py`
-   [ ] `test_motifs_online.py`
-   [x] `test_pairwise2.py`
-   [ ] `test_pairwise2_no_C.py`
-   [x] `test_pairwise_aligner.py`
-   [x] `test_phenotype.py`
-   [x] `test_phenotype_fit.py`
-   [ ] `test_phyml_tool.py`
-   [x] `test_prodoc.py`
-   [x] `test_prosite1.py`
-   [x] `test_prosite2.py`
-   [ ] `test_psw.py`
-   [x] `test_raxml_tool.py`
-   [x] `test_samtools_tool.py`
-   [x] `test_seq.py`
-   [x] `test_translate.py`

**Comment by svalqui** [comment]
I'll work on test_psw.py

**Comment by svalqui** [comment]
I'll work on test_phyml_tool.py

**Comment by svalqui** [comment]
I'll work on test_pairwise2_no_C.py

**Comment by svalqui** [comment]
I'll work on test_codonalign.py

**Comment by svalqui** [comment]
I'll work on test_cellosaurus.py

**Comment by svalqui** [comment]
I'll work on test_bgzf.py

**Comment by svalqui** [comment]
I'll work on test_Cluster.py

**Comment by svalqui** [comment]
I'll work on test_seq.py

**Comment by jvfe** [comment]
I'm new to contributing but I'll try working on these: `test_Prank_tool.py`,` test_Dialign_tool.py`, `test_Affy.py`, `test_NACCESS_tool.py`

**Comment by svalqui** [comment]
I'll work on test_CodonTable.py

**Comment by peterjc** [comment]
I've added this issue as a suggestion for the BCC2020 CoFest, since it is relatively straightforward to help with.

**Comment by jvfe** [comment]
I'll work on `test_SeqIO_Xdna.py`, `test_SeqIO_write.py` and `test_PDB_MMCIFParser.py`

**Comment by chrisbarnettster** [comment]
I'll work on:
- test_DSSP_tool.py 
- test_FSSP.py

**Comment by jvfe** [comment]
Not an extensive list, but from Tests/test_* I believe these are the ones remaining (checked if someone's already working on them):

- [ ]   Tests/test_AlignIO_convert.py
- [ ]   Tests/test_AlignInfo.py
- [ ]   Tests/test_BWA_tool.py
- [ ]   Tests/test_Blast_Record.py
- [ ]   Tests/test_Chi2.py
- [ ]   Tests/test_ClustalOmega_tool.py
- [ ]   Tests/test_Clustalw_tool.py
- [x]   Tests/test_Cluster.py
- [x]   Tests/test_CodonTable.py
- [ ]   Tests/test_ColorSpiral.py
- [ ]   Tests/test_Compass.py
- [ ]   Tests/test_Consensus.py
- [ ]   Tests/test_Crystal.py
- [x]   Tests/test_DSSP_tool.py
- [ ]   Tests/test_Emboss.py
- [x]   Tests/test_FSSP.py
- [ ]   Tests/test_Fasttree_tool.py
- [ ]   Tests/test_File.py
- [ ]   Tests/test_GenBank.py
- [ ]   Tests/test_HMMCasino.py
- [ ]   Tests/test_HMMGeneral.py
- [ ]   Tests/test_KEGG.py
- [ ]   Tests/test_KEGG_online.py
- [ ]   Tests/test_NMR.py
- [ ]   Tests/test_Nexus.py
- [ ]   Tests/test_PAML_codeml.py
- [ ]   Tests/test_PDB_MMCIF2Dict.py
- [ ]   Tests/test_Pathway.py
- [ ]   Tests/test_Phd.py
- [ ]   Tests/test_PopGen_GenePop.py
- [ ]   Tests/test_ProtParam.py
- [ ]   Tests/test_SeqIO.py
- [x]   Tests/test_SeqIO_Xdna.py
- [ ]   Tests/test_SeqIO_features.py
- [ ]   Tests/test_SeqIO_index.py
- [x]   Tests/test_SeqIO_write.py
- [ ]   Tests/test_Seq_objs.py
- [ ]   Tests/test_SwissProt.py
- [ ]   Tests/test_UniGene.py
- [ ]   Tests/test_Uniprot.py
- [ ]   Tests/test_align.py
- [ ]   Tests/test_align_substitution_matrices.py
- [x]   Tests/test_codonalign.py
- [ ]   Tests/test_pairwise2.py
- [ ]   Tests/test_pairwise_aligner.py

**Comment by padix-key** [comment]
I will work on `Tests/test_GenBank.py`.

**Comment by peterjc** [comment]
The end is in sight. Some will be tiny changes as the file was previously black formatted but some minor edits were done which didn't quite follow the style (and we missed that because it isn't yet enforced on the Tests/ folder).

**Comment by peterjc** [comment]
If you want to do a few related files together on a single pull request, that's fine too. e.g. ``Tests/test_SeqIO_*.py``

**Comment by mariecrane** [comment]
I'll work on Tests/test_Uniprot.py

**Comment by jvfe** [comment]
I'll work on `test_Emboss.py` and `test_Clustalw_tool.py`, fixing what was mentioned in #2835 as well.

**Comment by mariecrane** [comment]
I can work on Tests/test_KEGG.py

**Comment by jvfe** [comment]
I'll work on `test_HMM*`

**Comment by jvfe** [comment]
I'll work with the remaining `test_*_tool.py`

**Comment by jvfe** [comment]
I'll work on the remaining `test_Co*.py`

**Comment by jvfe** [comment]
I'll work on `test_pairwise*.py` and `test_AlignI*.py`

**Comment by jvfe** [comment]
I'll work on `test_P*.py`

**Comment by jvfe** [comment]
I'll work on `test_align*.py`

**Comment by jvfe** [comment]
I'll work on `test_Seq*.py`

**Comment by jvfe** [comment]
I'll work on `Tests/test_SwissProt.py` and `Tests/test_UniGene.py`, they also require changing assert statements, as specified in #2835

**Comment by peterjc** [comment]
@jvfe the final pull request to close this will be to update the flake8 configuration to stop ignoring black: https://github.com/biopython/biopython/blob/master/.flake8 setting ``per-file-ignores =``

And make a slight edit to https://github.com/biopython/biopython/blob/master/CONTRIBUTING.rst

And it would deserve mention in https://github.com/biopython/biopython/blob/master/NEWS.rst

**Comment by peterjc** [comment]
Almost there...

```
$ cd Tests/
$ black --check .
would reformat /Users/xxx/repositories/biopython/Tests/requires_wise.py
would reformat /Users/xxx/repositories/biopython/Tests/requires_internet.py
would reformat /Users/xxx/repositories/biopython/Tests/search_tests_common.py
would reformat /Users/xxx/repositories/biopython/Tests/PAML/gen_results.py
would reformat /Users/xxx/repositories/biopython/Tests/seq_tests_common.py
would reformat /Users/xxx/repositories/biopython/Tests/pairwise2_testCases.py
would reformat /Users/xxx/repositories/biopython/Tests/test_Seq_objs.py
```

**Comment by jvfe** [comment]
> Almost there...
> 
> ```
> $ cd Tests/
> $ black --check .
> would reformat /Users/xxx/repositories/biopython/Tests/requires_wise.py
> would reformat /Users/xxx/repositories/biopython/Tests/requires_internet.py
> would reformat /Users/xxx/repositories/biopython/Tests/search_tests_common.py
> would reformat /Users/xxx/repositories/biopython/Tests/PAML/gen_results.py
> would reformat /Users/xxx/repositories/biopython/Tests/seq_tests_common.py
> would reformat /Users/xxx/repositories/biopython/Tests/pairwise2_testCases.py
> would reformat /Users/xxx/repositories/biopython/Tests/test_Seq_objs.py
> ```

Oh, must've missed those cause I checked test_*. I'll do another PR for them.

**Comment by peterjc** [comment]
And we're done - a real team effort, thank you to everyone including all the BCC2020 CoFest participants who helped with the final push.
